### DISCUSSION TITLE
Requesting support for EEP D02-05-02

### DISCUSSION THREAD

**Discussion by OlwinFroon** [mak-gitdev/HA_enoceanmqtt#105] [Request]
Greetings!
This would be for my 5 Becker blinds controls.

I've checked the official EEP documentation and it looks like D02-05-02 is just a subset of D02-05-00. There are just a few minor differences:

CMD 1: Go to position and angle
- 4 bits @ offset 24: 
00: values 0 (channel 1) .. 3 (channel 4), 15 (all channels)
02: values 0 (channel 1), rest is "reserved"

CMD 2: Stop
- 4 bits @ offset 0: 
00: values 0 (channel 1) .. 3 (channel 4), 15 (all channels)
02: values 0 (channel 1), rest is "reserved"

CMD 3: Query Position and Angle
- 4 bits @ offset 0: 
00: values 0 (channel 1) .. 3 (channel 4), 15 (all channels)
02: values 0 (channel 1), rest is "reserved"

CMD 4: Reply Position and Angle
- 3 bits @ offset 21:
00: 0 (no lock), 1 (blockage), 2 (alarm)
00: 0 (no lock), 1 (blockage), rest is "reserved"
- 4 bits @ offset 24: 
00: values 0 (channel 1) .. 3 (channel 4), 15 (all channels)
02: values 0 (channel 1), rest is "reserved"

CMD 5: Set parameters
Doesn't exist in 02.

Thank you!

**Comment by mak-gitdev** [comment]
Hi @OlwinFroon,

I will have a [look](https://github.com/users/mak-gitdev/projects/1/views/1?pane=issue&itemId=47290789) 👍

**Comment by OlwinFroon** [comment]
Thank you, much appreciated.

Also, pairing a remote to a drive is covered by the documentation, but I didn't find anything about how to pair a shutter drive to the system.

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon, 

> Also, pairing a remote to a drive is covered by the documentation, but I didn't find anything about how to pair a shutter drive to the system.

Sorry I didn't get what you meant.
Are you looking for how to pair your D2-05-02 to Home Assistant ?

**Reply by OlwinFroon** [reply]
Yep.

**Reply by mak-gitdev** [reply]
@OlwinFroon,

This is explained [here](https://github.com/mak-gitdev/HA_enoceanmqtt/wiki/Pairing-your-device#procedure) in the wiki.
Your device has to be put in learning mode and for that, you will have to read its user manual to know how to do it.
Then, when in learning mode, the device will wait for anyone to reply with the corresponding pairing telegram.
Note that depending on the device type, the device may send pairing request telegrams while waiting for a reply. Your D2-05-02 falls into this category.

Then, once your device is in pairing mode, simply follow the procedure described in the link shared above to send that pairing telegram to complete the learn process.

**Reply by OlwinFroon** [reply]
Okay, I tried that. Power cycling is the only way to put these shutter drives in pairing mode for three minutes.

So I did that and switched ENOCEAN_LEARN on. Nothing happened.

According to the logs, the drive sends
```
2023-12-13 15:46:19,380 INFO: received: 05:19:79:63->FF:FF:FF:FF (-88 dBm): 0x01
['0xd1', '0x6', '0xc0', '0x0', '0x0', '0x0', '0x0', '0x7', '0x0', 
'0x5', '0x19', '0x79', '0x63', '0x81'] 
['0x0', '0xff', '0xff', '0xff', '0xff', '0x58', '0x0'] OrderedDict()
```
on power-up, followed by "message not interpretable".

If I'm not mistatken, that's a RORG D1 message ... which isn't defined by the EnOcean Alliance, right?

Immediately after that, the drive sends

```
2023-12-13 15:46:20,775 INFO: received: 05:19:79:63->FF:FF:FF:FF (-91 dBm): 0x01 
['0xd2', '0x0', '0x7f', '0x0', '0x4', 
'0x5', '0x19', '0x79', '0x63', '0x81'] 
['0x0', '0xff', '0xff', '0xff', '0xff', '0x5b', '0x0'] OrderedDict()

```
which would be the current shutter state (0%, unknown angle).

3 minutes later (end of pairing mode):

```
2023-12-13 15:49:21,534 INFO: received: 05:19:79:63->FF:FF:FF:FF (-89 dBm): 0x01 
['0xd1', '0x6', '0xc0', '0x0', '0x0', '0x0', '0x0', '0x7', '0x1', 
'0x5', '0x19', '0x79', '0x63', '0x81']
['0x0', '0xff', '0xff', '0xff', '0xff', '0x59', '0x0'] OrderedDict()

```
again followed by "message not interpretable".

:(

**Reply by OlwinFroon** [reply]
So, the D1 telegram might be manufacturer code for "install mode on/off".
It doesn't send anything else, let alone something that looks like a pairing request.
The docs don't say anything about pairing, except with new wall switches.

I guess I need to live with virtual F6-02-01 switches, those work at least. 
Thank you for that :)

**Reply by mak-gitdev** [reply]
@OlwinFroon,

It's normal that you get `message not interpretable`. It is because D2-05-02 is not yet supported.

D1 telegram are MSC (Manufacturer Specific Communication). This means that the decoding of such telegram is manufacturer-dependent and mostly not public. The manufacturer in this case seems to be `BECKER-ANTRIEBE` (ID = 0x06C).

For pairing, D2 devices normally use D4 telegrams (UTE - Universal TEach in).
Hence when the device is in learning mode, you should see D4 telegrams in the logs. Otherwise, it will be very difficult to pair with your D2-05-02 as it may require reverse-engineering ...

**Comment by OlwinFroon** [comment]
Yep, those are Becker drives, and they don't send D4 telegrams.

I didn't find anything about pairing in the [docs](https://www.manualslib.com/manual/2012366/Becker-Enocean-P5-16.html?page=20#manual). 
Best shot would be the next level of configuration: something called "Remote Commissioning"...

**Reply by mak-gitdev** [reply]
Yes aware of that and a very interesting feature.
It is somewhere in the "non-official" todo list (my head :joy:) but I am afraid it won't be anytime soon unfortunately.

At least you can use virtuals, so that's a good starting point.

By the way, regarding virtuals, starting from 0.1.28, I added virtual switches in addition to the existing virtual select.
Which one are you using ? Is it easy to use ? Do you have any comment/proposal to improve these ?

Thanks.

**Reply by OlwinFroon** [reply]
I tried pairing a virtual switches to one of the drives, and that went nicely. 
Meaning, the drive gave the response that was expected after pairing a new switch.
However, I couldn't quite get the hang of how it was supposed to work.

Per intuition I thought that switching AI on and then off would be the equivalent to tapping a wall switch's AI button and close the shutter. 
For some reason, it did not.
Setting the selector to "Action AI" and then to "none" worked, though.

I was about to peruse the communication logs to find out why, but then HA decided to auto-update to e2m version  0.1.28-1.
It crashed ever since, no log entries whatsoever and no idea. 

My wife wasn't too thrilled about the bedroom wall switch not working anymore :o)

This morning it occured to me it might be my custom mappings and EEP.

I had jury-rigged D2-05-02 support by copying and renaming the D2-05-00 section and jeez, editing those files is a major PITA.

Anyway, I disabled them for the time being, wall switch and sensors are working again.

Unfortunately, I won't have time to look into this again for the next 4 hrs...

**Comment by mak-gitdev** [comment]
Hi @OlwinFroon,

> I tried pairing a virtual switches to one of the drives, and that went nicely.

Note that the virtual switches represent the exact same device as the select, it is just a different UI to control that virtual device.
What I mean is that that drive is already controlled by the select entity, then no need to teach-in again the virtual switches (AI, AO, BI, BO).

<blockquote>
However, I couldn't quite get the hang of how it was supposed to work.

Per intuition I thought that switching AI on and then off would be the equivalent to tapping a wall switch's AI button and close the shutter.
For some reason, it did not.
</blockquote>

Yes it doesn't work like that.
You first have to configure the switches to the state you want and then press the `send` button.
I did that because of double press. For example, you may be required to press both `AI` and `BO` at the same time.
In this case, sending the command directly when the switch is activated doesn't work as it will first send for example `AI` alone when `AI` is switched, then `AI+BO` only when `BO` is finally switched after.
Do you have another idea of a more user-friendly mean to handle this case ?   

> My wife wasn't too thrilled about the bedroom wall switch not working anymore :o)

😅 Sorry for that.
Note that if this wall switch is used to control another EnOcean device, then you may think about pairing it directly to that device instead of doing it through HA. This way, if HA is KO, you can still control your device.
If it doesn't control an EnOcean device, then you have no choice unfortunately.

> I had jury-rigged D2-05-02 support by copying and renaming the D2-05-00 section and jeez, editing those files is a major PITA.

Yes I know,
I can't do anything for the XML one as this is a requirement from the Python EnOcean library.
For the mapping.yaml file, I am the only one culprit but is not that difficult once you start to be familiar with YAML.

**Reply by OlwinFroon** [reply]
Sorry, long day.

**Virtual switches ...:**
I started with V1.28, so I saw those switches right away, and I figured they were more flexible then the select, they just didn't work as I expected.

**... and how they work:**
Also, keeping an eye on the log, I figured the switch settings don't do anything before I hit send. Again, I don't know what went wrong there but meanwhile I'm tempted to blame it on the drive.
I found it doesn't respond to all telegrams, sometimes it just ignores them.
The switch approach seems to be the most flexible one. 
The most common use of the virtuals would be mimicking an actual wall switch, though, I guess. 
So a button with tap/release action like a real wall switch (you click it, it sends AI, you release it, it sends "released") would be nice. Not sure the HA UI can do that, though...

**Failing bedroom switch:**
I haven't that many EnOcean devices, just those 5 shutter drives, hand held switches for those, a handful of smart window handles (because they are so much nicer than battery powered door sensors) for heating and security reasons, and there's a bunch of wire- and batteryless wall switches (beats chopping up the wall for a power cable and beats battery powered switches, too) which act as light switches to control Zigbee lights via HA.
So except for the shutters, no switches are directly paired with EnOcean actors. 
I also have a few Zigbee remotes as a fallback, and HA on the phone works, too. 
My wife isn't open to the nerd approach, though, and prefers a wall switch to work. Can't exactly blame her for that :)

**EEP and mapping files:**
It's not the file structure per se, I don't have a problem with YAML or XML.
My issue with EEP.xml and mapping.yaml: they are huge and there's no easy way to find a specific entry.
Thankfully they are sorted alphanumerically, but you still have to leaf through them and try to keep track of RORG, func and type to find the entry you need, which is especially hard if the RORG row is like 15 pages up. And I'm using a 27" portrait display for stuff like that.
An XML attribute / YAML remark line containing the full EEP number prior to a device entry would help a LOT. 
Finding beats leafing by far :)

Whatever, I'll try and squeeze an hour or 3 off my schedule tomorrow, to track down what went wrong with those v-switches. Stay tuned.

**Reply by OlwinFroon** [reply]
Meh. Another looong day.

On the bright side: My contact sent me the complete tech specs of their EnOcean interface, including a command to directly pass low level commands to the drive's Modbus controller.

On a lesser bright side: pairing with HA and basically anything useful requires ReCOM.
And you don't do it by telling the drive to send pairing requests, I might say.
The only way is to edit HA's EnOcean address into the drive's outbound link table 🙄 

Well, that reduces the issue at hand to merely adding D2-05-02 support to the EEP and mapping files.
Yay!

Re v-switches: I'll be swamped until monday and won't be able to test those virtuals before tuesday. No promises, though ...

**Comment by OlwinFroon** [comment]
Okay, time for a bit of recollection and current status:

Initially after installing E2M, I edited the default EEP and mapping files to rig them for D2-05-02 by simply copying the D2-05-00 stuff. 
That worked perfectly until the update to 0.1.28-1, then it didn't.
E2M got stuck somewhere. 
The log went up to 
`Starting EnOceanMQTT...`
and then listed a bunch of services that stopped (sorry, I didn't copy the log).
End of action.
Funny enough, the Add-On interface still reported it as running (green dot).

I tried to restart, it didn't work until I removed the custom EEP.xml.

No idea what went wrong there, and I'm not exactly fond of the EEP monster, so I just removed the custom config and will wait for the official D2-05-02 support.

Meanwhile, I redefined my shutters as D2-05-00 (my OCD spidey senses are punishing me for that), restarted E2M and now I can see my shutter drives' current states.
Given the fact that they can be paired with HA via ReCOM only, a full integration seems to be far beyond the horizon anyway.
So this is as good as it gets, for now.

Next up I created a bunch of virtual switches to pair with the real drives, in order to gain SOME control over them, and it worked like a charm.
I even got them to work during at some point. 
Beats me how I did it, because the next day they didn't work anymore. 
Might've been the 0.1.28-1 update as well, might be because my wife asked me nicely to remove the "table full of tech crap" out of the living room, no idea.

So I started testing those virtual switches (see [here](https://github.com/mak-gitdev/HA_enoceanmqtt/discussions/108)) and found that the action select mode produces proper telegrams while the switch-and-send approach doesn't.

Right now the drives aren't responding to those, at all. 

Might be a range issue, now, since I had to rebuild my test rig in my man cave.

I'm gonna go find a repeater and put this on the backburner until I do.

Why can't anything be easy? Sigh... :(

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon,

Thanks for your feedback.

**EEP and mapping files:**
Ok I get your problem.
Depending on your editor, you should be able to fold/unfold the sections. This greatly simplify browsing those files.
But anyway I will add comments as you correctly suggested 👍 

**Virtual switches:**

<blockquote>
So a button with tap/release action like a real wall switch (you click it, it sends AI, you release it, it sends "released") would be nice. Not sure the HA UI can do that, though...
</blockquote>
Quite sure this can be done with some custom cards, but I am not a frontend expert.

**ReCom:**
It seems that you can achieve ReCom with DolphinStudio.
So if you have a Windows PC, you can have a look at this [link](https://www.enocean.com/wp-content/uploads/redaktion/support/dolphin-view/DolphinView/mod_remotecom.htm).
If you try it, depending on your feedback, I could try to add it if not too difficult.

**Support for D2-05-02:**
Please use the files below as custom EEP and mapping files.
[EEP.xml.zip](https://github.com/mak-gitdev/HA_enoceanmqtt/files/13714897/EEP.xml.zip)
[mapping.yaml.zip](https://github.com/mak-gitdev/HA_enoceanmqtt/files/13714900/mapping.yaml.zip)
There are two cover entities.
If `cover` entity seems to send or report inverted positions, then use the `cover_2` entity.
See #94 for more details about that.

**Issues:**
Even the select entity doesn't work ? That would be strange and unexpected.
Could you check that you have entities ending with `_2`. If so and if the select entity is still working, then this might be the issue.
In that case, the mapping.yaml file above should also fix that.
So may be not a range issue ...
Otherwise, please share your log file (with debug activated). The log file defaults to `/config/enoceanmqtt.log`.

**Reply by OlwinFroon** [reply]
Thanks for those files. Unfortunately, I have the same problem like before: using the custom EEP.xml doesn't agree with E2M:

```
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
Preparing to start...
Retrieved devices file: /config/enoceanmqtt/enocean.devices
use_dev_name_in_entity is USER-DEFINED (false)
Installing custom EEP.xml ...
cp: '/config/enoceanmqtt/EEP.xml' and '/config/enoceanmqtt/EEP.xml' are the same file
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped
```

That's the complete log, debug mode an all.
And on top of that, the add-on reports as running (green dot).

Removing the custom EEP results in E2M starting and running as expected.

The mapping file alone doesn't cause any trouble.

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon,

I just saw your message.

Ok I see.
The issue is in the addon file [run.sh](https://github.com/mak-gitdev/HA_enoceanmqtt-addon/blob/master/addon-dev/run.sh).
Line 123 should be:
```sh
   EEP_FILE_LOCATION=$(find / -path /config -prune -o -name 'EEP.xml' -print -quit 2>/dev/null)
```

If you are comfortable with Linux, VI file editor and more or less Docker, you can follow this [wiki](https://github.com/mak-gitdev/HA_enoceanmqtt/wiki/Contributing#testing-using-the-addon-or-docker) to modify the addon files and test both the fix and the support for D2-05-02.

**Reply by OlwinFroon** [reply]
Ugh. I'd love to, but this is a bit too deep for the old school tech afficionado that I am.

I'm comfy-ish with Linux and VI, but not at all with Docker. 
And I have no idea how to gain root access to my off-the-shelf Odroid N2 with HAOS.

Hints are welcome.

**Comment by OlwinFroon** [comment]
Hrm. Note to self: try the obvious next time.

So plugging in a keyboard and display worked. Kinda. 
I guess you're familiar with the downsides of working the console with a non-US keyboard.
Boy, that's a real pain in the posterior!

However, I edited run.sh as requested. No success, even after restarting everything.

I gather the problem was me putting EEP.xml in a separate directory instead of adding to the clutter /config, right?
Tried that (after the edit), doesn't work either.

Either the fix didn't work or I screwed up the edit.

I need to calm down from the annoying keyboard disaster before I take another look...

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon,

So you've been able to get access to the addon docker container, right ? Perfect.
> No success, even after restarting everything.

Just to clarify, to restart the addon with your changes, you have to run the command indicated in the wiki:
```sh
docker restart $(docker ps | grep ha_enoceanmqtt | cut -d" " -f 1)
```
If you restart the addon using the restart button of the addon, your changes are lost.

**Reply by OlwinFroon** [reply]
Yeah, I did the docker restart thingy and afterwards I was assaulted by a bunch of phone calls.
Bleh. Lots of distraction from the important stuff (= this).

Alright, I'll try again.

**Reply by OlwinFroon** [reply]
Um. I guess that's a bit harder than anticipated.

If I put my custom EEP in the configuration, E2M stops right away and I can't access the container.

If I don't, I CAN access the container to try and fix run.sh, but it won't be able to see the effect until I put my custom EEP in the configuration again, 
which would require a restart,
which would reset the run.sh again...?

**Reply by mak-gitdev** [reply]
@OlwinFroon, 

If it stops, it means that there is an error.
Could you share the log you get in this case ?

**Comment by OlwinFroon** [comment]
This is weird. I think there's more wrong with the custom configuration.

Back to square one: I set the configuration to 
- custom EEP and mapping  (those you sent me with D2-05-02 support) configured in directory /config/enoceanmqtt
- debug on
- log_packets on

I started E2M, the live log came up like the one posted above (services start, services stop).
There is no log file.
E2M stops eventually (takes a few seconds).

Then I put the new EEP.xml in /config directory, too.
The config entry for custom EEP still points to /config/enoceanmqtt/EEP.xml.
I started E2M again and, weird enough, it does NOT stop now, but works, as if I didn't have a custom EEP configured.

The log:

```
2023-12-20 16:22:11,947 INFO: Logging to file: /config/enoceanmqtt/enoceanmqtt.log
2023-12-20 16:22:11,948 INFO: Loading config file /data/enoceanmqtt.conf
:
sensor creation
:
2023-12-20 16:22:11,954 DEBUG: Global config: {'enocean_port': '/dev/ttyUSB1', 'log_packets': 'true', 'overlay': 'HA', 'db_file': '/data/enoceanmqtt_db.json', 'mapping_file': '/config/enoceanmqtt/mapping.yaml', 'ha_dev_name_in_entity': 'false', 'mqtt_discovery_prefix': 'homeassistant/', 'mqtt_host': 'core-mosquitto', 'mqtt_port': '1883', 'mqtt_client_id': 'enocean_gateway', 'mqtt_keepalive': '60', 'mqtt_prefix': 'enoceanmqtt/', 'mqtt_user': 'addons', 'mqtt_pwd': '*****', 'mqtt_debug': 'true'}
2023-12-20 16:22:12,008 INFO: Selected overlay : Home Assistant
2023-12-20 16:22:12,826 INFO: Mapping file correctly read: /config/enoceanmqtt/mapping.yaml
2023-12-20 16:22:12,827 INFO: Device database /data/enoceanmqtt_db.json correctly read/created
2023-12-20 16:22:12,827 INFO: Authenticating: addons
2023-12-20 16:22:12,828 DEBUG: Connecting to host core-mosquitto, port 1883, keepalive 60
2023-12-20 16:22:12,831 INFO: SerialCommunicator started
2023-12-20 16:22:12,832 INFO: Auto Teach-in is disabled
2023-12-20 16:22:12,832 WARNING: Replacing Packet.optional with default value.
2023-12-20 16:22:12,842 DEBUG: Sending CONNECT (u1, p1, wr0, wq0, wf0, c1, k60) client_id=b'enocean_gateway'
2023-12-20 16:22:12,932 INFO: Sending packet
2023-12-20 16:22:12,933 DEBUG: 0x05 ['0x8'] [] OrderedDict()
2023-12-20 16:22:13,035 DEBUG: 0x02 ['0x0', '0xff', '0xaa', '0xb7', '0x80'] ['0xa'] OrderedDict()
2023-12-20 16:22:13,036 INFO: got response packet: OK
2023-12-20 16:22:13,216 DEBUG: Received CONNACK (0, 0)
2023-12-20 16:22:13,217 INFO: Succesfully connected to MQTT broker.
```
followed by lots of other unsuspicious stuff and then, when I interacted with one of my shutters
```
2023-12-20 16:36:50,833 WARNING: Cannot find rorg 0xd2 func 0x5 type 0x2 in EEP!
```
which indicates that E2M doesn't use either of the custom EEP files.

FYI: I defined 3 of my drives as D2-05-00 and the other two as D2-05-02 for this test.



Okay, now I have a situation where the container still works with the configuration I want to use.
I guess I can now try and edit that shell script again ...

**Comment by OlwinFroon** [comment]
I did. It didn't work. Same as before, docker restart resulted in a stopped E2M with empty log file.
This is tedious and no fun at all, especially with that console keymap crap.

Can I catch you on discord?

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon,

Sorry just saw you message.
I think I have an account, but I don't really use it.
My ID there is `mak-dev`.

Regarding Docker, first of all thanks for your tests, but I don't want you to loose your mind ( :sweat_smile: ) so you can stop your tests  :slightly_smiling_face: .

**Comment by OlwinFroon** [comment]
My mind thanks you 😅 

I chimed in on a [year old thread](https://github.com/home-assistant/operating-system/issues/1775) about the console keymap issue, perhaps somebody will take pity on us non-americans.

So, no discord. Slack, perhaps?

I guess we save the custom EEP problem for later, I'll make a spin-off issue report for that.

Regarding the issue at hand (missing D2-05-02 coverage), I'll just wait for the next E2M version and hope you add it anyway, without the need for custom EEP files.

Once that works, I'll explore the curious [action vs. switches thing](https://github.com/mak-gitdev/HA_enoceanmqtt/discussions/108) and perhaps take a peek at the ReMAN/ReCOM business.

**Comment by OlwinFroon** [comment]
Thank you very much for your help so far!

Back to testing: I configured all my shutter drives as D2-05-02 using your test EEP and mapping files. The drives show up in HA just fine, no error messages. 
They show "open" or "closed" and the shutters´ current positions. 
For some of them it's the first cover sensor, for others it´s cover2. 
Not sure why, I guess it depends on the orientation of the drive motor (left or right)
Doesn´t matter. 

**And here comes issue number 1:**

One of my drives kept acting up during installation, it refused to self-teach the lower boundary. 
It took some fiddling on behalf of the guy who installed them, but now it works. 
Only problem: the "closed" position is not at 0, but at 3.
So even if this one is fully closed, the icon says "open" and the interface would allow to hit the "close" button.
I could of course have a bunch of template sensors telling me the shutter states (open, moving, closed, unknown) instead of the device sensors, but perhaps there's a more elegant way to solve this.


Back to topic: We know I can't pair my drives with HA without ReCOM (and I won't play with that anytime soon), so the controls on the device page won't work.
For further testing, I paired two of the drives with virtual switches (I just LOVE these. Thank you!).

**... which brings us to issue number 2:**

When I send an "OPEN" or "CLOSE" command to a drive, it immediately responds with 
`['0xd2', '0x7f', '0x7f', '0x0', '0x4', '0xa1', '0xa2', '0xa3', '0xa4', '0x80']`
which means "position unknown" and actually means "either we had a power cycle or the drive is in motion".
As soon as the motor stops, it sends a new telegram with the real position, which is usually 0 or 0x64 = 100(% open).

The position attribute in HA, however, never shows 0x7f, 127 or "unknown". It always translates 0x7f to "100", which makes it impossible to say if the shutters are really open (100) or in motion, ... or make a template sensor for the current state of a shutter (see above).


Issue number 3 is not directly related to E2M: One of the drives seems to have a poorly adjusted antenna and won´t receive all commands right away.
I very few cases it does, but mostly I need to send twice. Even three times once in a while.

In order to end up with unattended shutter control, I'll try and fix this with a repeater and/or a blueprint that listens for a response from the drive and tries again if there is none.
No need to bother you with this :)

**Comment by mak-gitdev** [comment]
Hi @OlwinFroon,

First of all, thanks for your feedback :+1: 

**Issue 1**:

<blockquote>
Only problem: the "closed" position is not at 0, but at 3.
So even if this one is fully closed, the icon says "open" and the interface would allow to hit the "close" button.
I could of course have a bunch of template sensors telling me the shutter states (open, moving, closed, unknown) instead of the device sensors, but perhaps there's a more elegant way to solve this.
</blockquote>
For now, unfortunately I have no other solution than using a template sensor to handle this case.

**Issue 2**:

Yes you're right, the mapping as currently done doesn't correctly handle the position value 127.
I think we can ignore this value when it is received instead of interpreting it as 100%. Would this be OK for you or do you have a better idea ?

**Reply by OlwinFroon** [reply]
Hi @mak-gitdev,

Issue #1: 
I'd rather use the UI's fancy cover tiles, but they'd need to be a lot more customizable for my special situation. 
So, yeah, I'll be stuck with a template sensor and a bunch of buttons instead.
I wonder if device file parameters couldn't solve this (and the cover/cover2 thing):
`closedposition=0`
`openposition=100`

... or something like that.

Issue #2:
Right now, I get false values and can't tell what's happening.
I'd rather have the real raw values including the 127 instead of no or trimmed ones.

After a power cycle, 127 means the drive gave a lifesign and the current position is unknown.
After a known position, 127 means the drive is moving.
My template sensor could handle that nicely.

I don't know it the 127 would mess up the pretty cover tiles but can't use them anyway so I don't care much :D

A parameter in the device file could fix that, though:
`ignoreinvalidpositions=false`
(default value would be `true`).

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon, 

**Issue 1**

You're right, you can use a custom mapping file and create a cover3 entity based on one of the existing covers.

**Issue 2**

I think ignoring that value will work.
I will later share a mapping for that if you want.

**Device options**

I have to figure out how I can easily add such options.

**Reply by OlwinFroon** [reply]
Hi @mak-gitdev,

#1: i'll give it a try,

#2: yes please,

and, as always, thanks for your work :)

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon,

For #2, you can replace:
```yaml
                 position_template: "{{ value_json.POS }}"
                 tilt_status_template: "{{ value_json.ANG }}"
```
By:
```yaml
                 position_template: >-
                   {% if (value_json.POS <= 100) and (value_json.POS >= 0) %}
                     {{ value_json.POS }}
                   {% else %}
                     {{ state_attr(entity_id,"current_position")|int(default=0) }}
                   {% endif %}
                 tilt_status_template: >-
                   {% if (value_json.ANG <= 100) and (value_json.ANG >= 0) %}
                     {{ value_json.ANG }}
                   {% else %}
                     {{ state_attr(entity_id,"current_tilt_position")|int(default=0) }}
                   {% endif %}
```

<details><summary>Full mapping</summary>

```yaml
            - component: "cover"
              name: "cover"
              config:
                 command_topic: "req"
                 payload_open: >-
                   {"CMD":"1","POS":"0","ANG":"127","REPO":"0","LOCK":"0","CHN":"0","send":"clear"}
                 payload_close: >-
                   {"CMD":"1","POS":"100","ANG":"127","REPO":"0","LOCK":"0","CHN":"0","send":"clear"}
                 payload_stop: >-
                   {"CMD":"2","CHN":"0","send":"clear"}
                 position_topic: "CMD4"
                 position_template: >-
                   {% if (value_json.POS <= 100) and (value_json.POS >= 0) %}
                     {{ value_json.POS }}
                   {% else %}
                     {{ state_attr(entity_id,"current_position")|int(default=0) }}
                   {% endif %}
                 set_position_topic: "req"
                 set_position_template: >-
                   {"CMD":"1","POS":"{{ position }}","ANG":"127","REPO":"0","LOCK":"0","CHN":"0","send":"clear"}
                 tilt_status_topic: "CMD4"
                 tilt_status_template: >-
                   {% if (value_json.ANG <= 100) and (value_json.ANG >= 0) %}
                     {{ value_json.ANG }}
                   {% else %}
                     {{ state_attr(entity_id,"current_tilt_position")|int(default=0) }}
                   {% endif %}
                 tilt_command_topic: "req"
                 tilt_command_template: >-
                   {"CMD":"1","POS":"127","ANG":"{{ tilt_position }}","REPO":"0","LOCK":"0","CHN":"0","send":"clear"}
            - component: "cover"
              name: "cover2"
              config:
                 command_topic: "req"
                 payload_open: >-
                   {"CMD":"1","POS":"0","ANG":"127","REPO":"0","LOCK":"0","CHN":"0","send":"clear"}
                 payload_close: >-
                   {"CMD":"1","POS":"100","ANG":"127","REPO":"0","LOCK":"0","CHN":"0","send":"clear"}
                 payload_stop: >-
                   {"CMD":"2","CHN":"0","send":"clear"}
                 position_topic: "CMD4"
                 position_open: 0
                 position_closed: 100
                 position_template: >-
                   {% if (value_json.POS <= 100) and (value_json.POS >= 0) %}
                     {{ value_json.POS }}
                   {% else %}
                     {{ state_attr(entity_id,"current_position")|int(default=0) }}
                   {% endif %}
                 set_position_topic: "req"
                 set_position_template: >-
                   {"CMD":"1","POS":"{{ 100 - position }}","ANG":"127","REPO":"0","LOCK":"0","CHN":"0","send":"clear"}
                 tilt_status_topic: "CMD4"
                 tilt_status_template: >-
                   {% if (value_json.ANG <= 100) and (value_json.ANG >= 0) %}
                     {{ value_json.ANG }}
                   {% else %}
                     {{ state_attr(entity_id,"current_tilt_position")|int(default=0) }}
                   {% endif %}
                 tilt_command_topic: "req"
                 tilt_command_template: >-
                   {"CMD":"1","POS":"127","ANG":"{{ tilt_position }}","REPO":"0","LOCK":"0","CHN":"0","send":"clear"}
```
</details>

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon,

Did you have time to test the code above for your D2-05-02 ?

I am about to make a new release, do you have any remaining bugs ?
I would like this release to serve as basis for the first stable version so I would like it to fix as much bugs as possible.

Thanks.

**Reply by OlwinFroon** [reply]
I did, but it didn't seem to make much of a difference, so I reverted to the original version.

At this point, there are only a few minor things that are bothering me mildly:

1)
A few of my shutters are running on 'cover', the others on 'cover2'. 
I think there should be only one cover entity and it's behaviour should be controlled by parameters (direction and endpoints)  in the devices file.

2)
Two of my drives don't always accept commands right away for reasons I'm still investigating. 
Problem is, I don't get the 'position 127' responses, so it is hard to know if they did or not, so my automations can't tell. 
As a result, they have to wait a few seconds, check if the position has changed and possibly re-send the command.
Processing the '127' responses into a binary sensor ('unknown or moving') would solve this.

3)
My drives support the "angle" aspect through their EEP, but my shutters don't. 
The whole angle thing should be optional, disabling this via devices file parameter would do the trick.

But to be honest, all of the above are merely tweaks.
I'm quite happy that this works at all, and I can very well live with the current 'official' solution, which is simply the duplication of D2-05-00 into D2-05-02.

That is a solid basis to work with, the tweaks can wait.

I am not aware of any real bugs with D2-05-02 so far, so I guess this is safe to use for the first official stable version.

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon, 

Thanks for your feedback and I agree that those are minor issues that we can deal with later.

**Reply by TheFamousSpy** [reply]
Hi Guys,

just saw your discussion and I am really excited that support for D2-05-02 comes soon. I am also have some Becker shutters installed.
If I can contribute something, especially testing, I am happy to help.

@mak-gitdev, 
I hope you are adding a donation button to the contributing page ASAP because you really deserve it for your effort!

**Comment by OlwinFroon** [comment]
@mak-gitdev do you think it might be useful to add important information and peculiarities directly to the Wiki / Supported devices?

In this case: 

_D2-05-02 Becker shutter drives are not fully supported yet.
They will show up in E2M as cover devices, but they will appear read-only. The control buttons will not be functional.
Reason: Full support would require pairing the shutter drive to your EnOcean controller.
In case of these drives, that would require the Remote Management and Remote Commissioning features, but E2M doesn't support these yet.
Workaround: define a virtual button and pair it directly to the drive. That way you will be able to open and close your shutters via Home Assistant._

**Reply by mak-gitdev** [reply]
Hi @OlwinFroon,

I was also thinking about that since your comments on D2-03-0A.
So I will definetely do that and thanks for the text 👍
